### DISCUSSION TITLE
Lora example doesn't work?

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#989] [Getting started] [answered]
I decided to try the LoRa example (`examples/drivers/lora). When I run it on esp32 I get an error:
```
/data/main.js (26) # Break: main: get device: undefined variable!
```
`device` is not imported, is it a global (newbie here, obviously)? This demo is not listed in the readme for the drivers directory, so maybe I'm missing some prerequisite?

**Comment by phoddie** [comment]
Please provide the command line you used to build the example.

**Reply by tve** [reply]
```
> mcconfig -d -m -p esp32/nodemcu                       
Detecting the Python interpreter                                                  
Checking "python" ...                                                             
Python 3.10.8                                                                     
"python" has been detected                                                        
Not using an unsupported version of tool xtensa-esp32-elf found in PATH: esp-2021r
2-patch3-8.4.0.                                                                   
Python requirements from /home/sw/moddable/esp-idf/requirements.txt are satisfied.
# Reconfiguring ESP-IDF...                    
...
[100%] Built target xs_esp32.elf
[100%] Built target gen_project_binary
xs_esp32.bin binary size 0x5d380 bytes. Smallest app partition is 0x390000 bytes. 
0x332c80 bytes (90%) free.
[100%] Built target app_check_size
[100%] Built target app
esptool.py esp32 -p /dev/ttyUSB1 -b 921600 --before=default_reset --after=hard_res
et write_flash --flash_mode dio --flash_freq 80m --flash_size 4MB 0x1000 bootloade
r/bootloader.bin 0x10000 xs_esp32.bin 0x8000 partition_table/partition-table.bin
esptool.py v3.3.2
Serial port /dev/ttyUSB1
...
Leaving...
Hard resetting via RTS pin...
[100%] Built target flash
 ESP32
Running make in directory /home/sw/moddable/moddable/build/tmp/esp32/nodemcu/debug/lora/xsProj-esp32/build
Executing "make -j 18 flash"...
Done
```

![image](https://user-images.githubusercontent.com/39480/207794856-1f124bdf-4886-4e65-9240-193dd48748b7.png)

**Comment by mkellner** [comment, marked as answer]
I see your build target is `esp32/nodemcu`. That device does not support LoRa.

There is a build target `esp32/heltec_lora_32` located in `$MODDABLE/build/devices/esp32/targets/heltec_lora_32`. It is set up for the way that device is wired.

The global `device` is defined in `host/provider.js` in that directory. That particular build target includes a `device.peripheral.lora` object, which is necessary for this example.

**Reply by tve** [reply]
Yikes. Thanks for pointing that out.
I'm not using the heltec board, just my own PCB, so I thought a generic target would be appropriate.
So I guess option 1 is to just use the heltec target and live with the extra stuff and hope nothing creates a conflict, or option 2 dive into how I define my own build target?

**Reply by mkellner** [reply]
Disclaimer: I've not used any LoRa device myself. I give you general information that I am gleaning by looking at the target's manifest and provider. I don't know how generic LoRa is, or if there are special commands, etc. From the provider, I see that the heltec device uses SPI at 9 MHz and has three additional pins (other than SPI pins): `reset`, `interrupt` and `select`.

I am assuming your custom PCB is using the ST127x LoRa chipset or compatible. (confirmed with issue #990)

But even if it isn't, I think option 2 would be best here if you've got a custom board.

Copy the whole `heltec_lora_32` directory to, let's say, `custom`. Then your build target would be `esp32/custom`.

The heltec board looks like it has a display: `ssd1306`. You can excise that from the `manifest.json` file by removing the ssd1306 `include`, changing `"screen"` in the config object to just an empty string (like `touch` below it), and remove the `ssd1306` section (and the comma above it - don't leave dangling commas).

Also in the `manifest.json` file, change the defines for `i2c` and `spi` to match the GPIO pins you've got set up for those interfaces.

Then, onto the `host/provider.js` file. The `const device` has pin definitions for I2C, Serial, SPI, Analog. Change those for your device. If you're not using something (like say, Analog), you don't need to define a `default`.

Also in `provider.js`, you can remove the `backlight`, `displayDC` and `displaySelect`. If you've got an LED defined, you can specify it's pin in the `pin` section. The same with a button.

The example looks like it expects a button. That is defined as `peripheral.button.Flash`.
You can remove the `Backlight` section in `peripheral` if you don't have a screen.

In the `peripheral.lora`, you'd change the `pin` definitions for `reset`, `interrupt` and `select` as wired for your device.

**Reply by tve** [reply]
Thanks for the detailed reply!

**Reply by mkellner** [reply]
If you've got questions while creating your new target, please lets start a different discussion topic for that.

**Reply by tve** [reply]
Quick follow-up: the sx127x driver is now saying it recognizes the chip on my board, yay!

**Reply by nickthiru** [reply]
Hi @tve

I was wondering if you might share how you setup the example to work, please. (I am assuming you were trying to get the example to work with an `esp32/nodemcu`). I am stuck at the chip recognition stage, as seen in the following screenshot:

![image](https://user-images.githubusercontent.com/47299208/226251242-b53c6f5c-0973-4bc8-b249-caf5b2a7be6b.png)

I am assuming you would have updated the `defines` object in `build\devices\esp32\targets\heltec_lora_32\manifest.json` and `devices.SPI.default` and `defines.peripheral.lora.Default` constructor object in 'build\devices\esp32\targets\heltec_lora_32\host\provider.js'?

**Reply by tve** [reply]
What is your hardware configuration? Are you using the Heltec board? If not, what are you using, or how are things connected?
I have a board I call rfgw (RF gateway) and I created a board definition for it with the pin assignments it uses. Happy to help you get that together if that's what you need.

**Reply by nickthiru** [reply]
@tve Thanks for the prompt response. I am trying to get the `examples\drivers\lora\` example to work with `esp32/nodemcu`. Here's what I did.

1. I copied the `build\devices\esp32\targets\heltec_lora_32\` folder to `build\devices\esp32\targets\nodemcu_lora\`
2. I also wanted to keep the new target as close to `build\devices\esp32\targets\nodemcu_lora\`, which is why the new target folder also has the `setup-target.js`, which I did not modify.

![image](https://user-images.githubusercontent.com/47299208/226341193-c0231610-6a66-4627-b930-09f09230b7b3.png)

3. The new target's manifest.

![image](https://user-images.githubusercontent.com/47299208/226341829-382af147-8737-4e8f-b088-335dd873c67b.png)

4. The new target's provider. I added the `device.SPI.VSPI` property and set the VSPI (as opposed to HSPI) pins there.

![image](https://user-images.githubusercontent.com/47299208/226342581-98710f11-e392-4090-9ec2-db72ae60317f.png)

5. To the provider, I also added the property `device.peripheral.lora.VSPI' with a class value that was copied from `device.peripheral.lora.Default', but with the `spi` constructor object referencing the new VSPI object, and the `select` constructor object pin value set to the VSPI CS pin number.

![image](https://user-images.githubusercontent.com/47299208/226345609-aa847135-1b26-41ff-a399-93e2c6138e22.png)

6. In `examples\drivers\lora\main.js`, I updated the call to the lora initializer to the new `VSPI` property I created in the new target's provider.

![image](https://user-images.githubusercontent.com/47299208/226347037-9e9665af-c47c-4edc-9306-1bce5029f0f6.png)

7. Finally, I executed `mcconfig -d -m -p esp32/nodemcu_lora` from `examples\drivers\lora\`.

**Reply by tve** [reply]
Why are you creating all the duplication with VSPI instead of updating the SPI pins and being done with it?
I'm not seeing what's wrong, I assume you double-checked your wiring. One thing to try is to back down from 9Mhz SPI clock to 1Mhz or lower.

**Reply by nickthiru** [reply]
> Why are you creating all the duplication with VSPI instead of updating the SPI pins and being done with it?

I did this because I may do a PR down the road with the proposal that the `nodemcu_lora` target I created replace the `nodemcu` target. In this case, creating a separate configuration for the VSPI interface would reduce the chances of breaking any code that relies on the `nodemcu` target as it exists now i.e. the default using HSPI pins. 

> I'm not seeing what's wrong, I assume you double-checked your wiring. One thing to try is to back down from 9Mhz SPI clock to 1Mhz or lower.

Yes, that's what I thought to: check the wiring. Thanks for tip to check the clock speed; I will try that as well.

**Reply by phoddie** [reply]
I've not been following this in detail, so apologies if this is off-topic. In the SPI constructor options object, the port property is important. On ESP32, it corresponds to the ESP-IDF enumeration SPI1_HOST, SPI2_HOST, SPI3_HOST. The IDF makes this a bit confusing, but for an original ESP32 (not ESP32-S2, ESP32-S3, etc), setting port to 1 in the SPI constructor selects SPI2_HOST which is HSPI_HOST. If you want to use VSPI_HOST, the port needs to be set to SPI3_HOST which is 2. (You can see all this for yourself in ESP-IDF/components/hal/include/hal/spi_types.h.)

**Reply by nickthiru** [reply]
@phoddie Thanks for the insight. I was, in fact, wondering about the purpose of the `port` property.

**Comment by nickthiru** [comment]
Good thing I came across this (and #991); I was about to implement the SX1276 driver for my nodemcu! Thanks @tve for asking and @mkellner for the detailed answer.

**Comment by nickthiru** [comment]
Hi,

I'm wondering how it's possible for the lora example to work (esp. accessing the `device` object), even when selecting `esp32/heltec_lora_32` as the target, when its [manifest](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/lora/manifest.json) doesn't reference, neither directly nor indirectly, `modules\io\manifests\esp32\manifest.json`, which is where the `provider.js` modules are included (specifically, within the `platforms` object)?

**Reply by phoddie** [reply]
- esp32/heltec_lora_32 includes [$(MODDABLE)/modules/io/manifest.json](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/io/manifest.json)
- which includes [./manifests/esp32/manifest.json](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/io/manifests/esp32/manifest.json)
- which includes "$(BUILD)/devices/$(PLATFORM)/targets/$(SUBPLATFORM)/host/provider
- here "$(BUILD)" is `$MODDABLE/build`, "$(PLATFORM)" is `esp32` and "$(SUBPLATFORM)" is `heltec_lora_32`

**Reply by nickthiru** [reply]
@phoddie Thank you. Yes, I did notice the first one but I didn't see how that manifest was being included in the build process. If I execute `mcconfig` from `examples/drivers/lora/`, that example's manifest only includes `/examples/manifest_base.json`, and when I followed the chain _downwards_, there was no connection with `esp32/heltec_lora_32`'s manifest. I thought that the build target's manifest may be listed in the `platforms.includes` property of `modules\io\manifests\esp32\manifest.json` but there is only the `modules` property in the platforms object, which references the provider but not the manifest.

In relation to manifests, does the following...
![image](https://user-images.githubusercontent.com/47299208/225182122-21e9d671-cbda-4611-844d-dfdbb344e056.png)
...mean that it starts with manifests that are _deepest_ in the chain and then works its way _up_? So, the deepest would be `examples\manifest_base.json` and the top-most one would be the manifest of the `host` or `mod`.

**Reply by nickthiru** [reply]
> @phoddie Thank you. Yes, I did notice the first one but I didn't see how that manifest was being included in the build process. If I execute `mcconfig` from `examples/drivers/lora/`, that example's manifest only includes `/examples/manifest_base.json`, and when I followed the chain _downwards_, there was no connection with `esp32/heltec_lora_32`'s manifest. I thought that the build target's manifest may be listed in the `platforms.includes` property of `modules\io\manifests\esp32\manifest.json` but there is only the `modules` property in the platforms object, which references the provider but not the manifest.

Ok, I found the link in `examples\manifest_base.json`

![image](https://user-images.githubusercontent.com/47299208/225188663-ac628e80-e13d-42bb-899f-8b633320d2ab.png)

**Reply by phoddie** [reply]
Yes, exactly. That's why every host build eventually needs to include manifest_base.json, as that's what pulls in the core platform specific modules, etc.
